generic-ec/src/multiscalar.rs

```diff
@@ -0,0 +1,404 @@
+///    initially set to $\O$, i.e. $B_1 = B_2 = \dots = B_{15} = \O$
+/// 2. Sort points $P_{1..n}$ into buckets $B$: for each $1 \le i \le n$, set
+///    $B_{s_{i,j}} := B_{s_{i,j}} + P_i$
+/// 3. Compute $S = 1 B_1 + 2 B_2 + \dots + 15 B_{15}$. To do this efficiently:
+///    1. Set $S' := B_{15}$, $S := S'$
+///    2. Then, for each $m \in [14, 13, \dots, 1]$, do: \
+///       Set $S' := S' + B_m$, and $S := S + S'$
+/// 4. Set $Q := 16 Q + S$
+///
+/// Output $Q$
+///
+/// ## How it works
+/// Similarly to [`Straus`] algorithm, here we work with 2-dimention sum:
+///
+/// $$
+/// \begin{aligned}
+/// s_1 P_1 &=&& s_{1,0} P_1 &&+&& 16^1 s_{1,1} P_1 &&+ \dots +&& 16^{k-1} s_{1,k-1} P_1 \\\\
+///    \+   & &&        +    && &&             +    &&         &&                   +    \\\\
+/// s_2 P_2 &=&& s_{2,0} P_2 &&+&& 16^1 s_{2,1} P_2 &&+ \dots +&& 16^{k-1} s_{2,k-1} P_2 \\\\
+///    \+   & &&        +    && &&             +    &&         &&                   +    \\\\
+/// \vdots  & && \vdots      && && \vdots           &&         && \vdots                 \\\\
+///    \+   & &&        +    && &&             +    &&         &&                   +    \\\\
+/// s_n P_n &=&& s_{n,0} P_n &&+&& 16^1 s_{n,1} P_n &&+ \dots +&& 16^{k-1} s_{n,k-1} P_n
+/// \end{aligned}
+/// $$
+///
+/// We compute the sum column by column from right to left. The difference is how
+/// $s_{1,j} P_1 + s_{2,j} P_2 + \dots + s_{n,j} P_n$ is computed: Pippenger algorithm uses
+/// the buckets trick as described above.
+#[cfg(feature = "alloc")]
+pub struct Pippenger;
+
+#[cfg(feature = "alloc")]
+impl<E: Curve> MultiscalarMul<E> for Pippenger {
+    fn multiscalar_mul<S, P>(scalar_points: impl IntoIterator<Item = (S, P)>) -> Point<E>
+    where
+        S: AsRef<Scalar<E>>,
+        P: AsRef<Point<E>>,
+    {
+        let (mut scalars, points): (Vec<Radix16Iter<E>>, Vec<Point<E>>) = scalar_points
+            .into_iter()
+            .map(|(scalar, point)| (scalar.as_ref().as_radix16_be(), *point.as_ref()))
+            .unzip();
+        Self::mutliscalar_mul_inner(&mut scalars, &points)
+    }
+}
+
+#[cfg(feature = "alloc")]
+impl Pippenger {
+    /// Takes `scalars` represented in big-endian radix 16, `points`, and performs mulitscalar
+    /// multiplication, returns `sum = scalars[0] * points[0] + ... + scalars[n-1] * points[n-1]`
+    ///
+    /// Requires that `scalars.len() == points.len()`
+    fn mutliscalar_mul_inner<E: Curve>(
+        scalars: &mut [Radix16Iter<E>],
+        points: &[Point<E>],
+    ) -> Point<E> {
+        if scalars.is_empty() {
+            return Point::zero();
+        }
+
+        // `serialized_len` is amount of radix256-digits. We multiply it by 2 and get
+        // amount of radix16-digits
+        let num_digits = 2 * Scalar::<E>::serialized_len();
+
+        let mut result = Point::zero();
+        for i in 0..num_digits {
+            let mut buckets = [None; 15];
```

Comment:
Why not start with zeroes and get rid of matching below?

Reply:
Because I want to minimize additions as this is the point of the algorithm. Setting it to Point::zero() would add 1 extra addition per each bucket

Reply:
After thinking about it, you might be right, especially if points are CT; and if they are not CT it should be comparable

Reply:
I actually benchmarked it for n=150 and 100, and the difference is 8% in favour of branching. Ok